### ISSUE TITLE
Add MPI_IN_PLACE option to Alltoall

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#31]
**Originally by _traff_ on 2008-10-10 08:22:19 -0500**

---
# Add MPI_IN_PLACE option to AlltoallAuthor: Torsten Hoefler

Update: Jesper Larsson Träff, Torsten Hoefler
## Status

 Sep 8 2008 Straw Vote

For the idea of using IN_PLACE:

|| Yes || Maybe || No for 2.2 || Abstain ||
|| 24  || 0     ||  0         ||    1    ||

For the idea of adding new routines:

|| Yes || Maybe || No for 2.2 || Abstain ||
|| 5 || 25 || 0 || 2 ||
## Description

Allow `MPI_IN_PLACE` to `MPI_Alltoall` as a send (or receive) buffer argument, with the intention that data to be sent and received are taken from the same buffer. Thus, the effect of the _in-place_ option in this case will be that data in the buffer are replaced. The call can be seen as an analogue of `MPI_Sendrecv_replace`.

Since sendbuffers may likely be typed `const`, the proposal is that `MPI_IN_PLACE` can be specified as the `sendbuf` argument. In that case, the data to be sent must be placed in the receive buffer. Upon completion of the operation, they will have been replaced with the data from all other processes (in the usual rank order).

It could make sense for algorithms to know that **all** processes are in _replace_-mode, that is to require that if some process calls with `MPI_IN_PLACE`, then **all** must do so!

The idea could be extended to `MPI_Alltoallv`, although the requirements on the `count` vectors may be difficult for users
to deal with. Nevertheless, `MPI_IN_PLACE` makes perfect
sense here also, and a proposal is made below.

See also Ticket #113
## History

Several users (>2) asked me for it. It makes sense in memory constrained situations.

Should check the somewhat more involved/constrained operation discussed in Siegel's MADRE paper, EuroPVM/MPI 2008.
### Part A

Replace p. 155, line 46 by the following text:

The "in place" option for intracommunicators is specified by passing `MPI_IN_PLACE` to the argument `sendbuf` at _all_ processes. In such a case, `sendcount` and `sendtype` are ignored. The data to be sent is taken from the `recvbuf` and replaced by the received data. Data sent and received must have the same type map as specified by `recvcount` and `recvtype`.

-Rationale_:
For large `MPI_Alltoall` instances, allocating both send and receive buffers may consume too much memory. The "in place" option effectively halves the application memory consumption and is useful in situations where the data to be sent will not be used by the sending process after the `MPI_Alltoall` exchange (e.g., in parallel Fast Fourier Transforms).
(_End of rationale.*)

-Advice to Implementors_:
Users may opt to use the "in place" option in order to conserve memory. Quality MPI implementations should thus strive to minimize system buffering.
(_End of advice to implementors.*)
### Part B

P. 157, line 20, replace:

No "in place" option is supported.

by

The "in place" option for intracommunicators is specified by passing `MPI_IN_PLACE` to the argument `sendbuf` at _all_ processes. In such a case, `sendcounts`, `sdispls` and `sendtype` are ignored. The data to be sent is taken from the `recvbuf` and replaced by the received data. Data sent and received must have the same type map as specified by the `recvcounts` array and the `recvtype`, and is taken from the locations
of the receive buffer specified by `rdispls`.

-Advice to Users_:
Specifying the "in place" option (which must be given on all processes) implies that the same amount and type of data is sent and received between any two processes in the group of the communicator.
Different pairs of processes can exchange different amounts of data.
Users must ensure that `recvcounts[j]` and
`recvtype` on process `i` match `recvcounts[i]` and
`recvtype` on process `j`. This symmetric exchange can be useful in applications where the data to be sent will not be used by the sending process after the `MPI_Alltoallv` exchange. (_End of advice to users*.)

Replace p. 159, line 17 with:

Like for `MPI_Alltoallv`, the "in place" option for intracommunicators is specified by passing `MPI_IN_PLACE` to the argument `sendbuf` at _all_ processes. In such a case, `sendcounts`, `sdispls` and `sendtypes` are ignored. The data to be sent is taken from the `recvbuf` and replaced by the received data. Data sent and received must have the same type map as specified by the `recvcounts` and `recvtypes` arrays, and is taken from the locations
of the receive buffer specified by `rdispls`.
## Impact on Implementations

An implementation in terms of `MPI_Sendrecv_replace()` can trivially be done. Better solutions exist (of course), so efficient `MPI_Alltoall` implementations need to be changed. 
## Impact on Applications / Users

For `MPI_IN_PLACE` there is none, it is an addition (that some 
scenarios may benefit from). 
## Alternative Solutions

An alternative would be to have (in analogy with `MPI_Sendrecv_replace` a dedicated collective

```
#!c
MPI_Alltoall_replace(sendrecvbuf,...)
```

Deprecate MPI_Sendrecv_replace() and add the MPI_IN_PLACE option to MPI_Sendrecv() so that the symmetry in the standard is retained. But the Forum was widely opposed to this proposal (as it would require an additional branch in the performance-critical MPI_Sendrecv()).

htor: I don't think that we need the rationale below because this is already enforced for other All-operations, such as Allreduce and Allgather. {Ex}Scan is an exception here, see #94!

Rationale: "In order to stay symmetric, we propose to deprecate MPI_Sendrecv_replace() and add the MPI_IN_PLACE option to MPI_Sendrecv()."
## Entry for the Change Log

For Part A:

Section 5.8 on page 155.
"In place" option is added to MPI_ALLTOALL, MPI_ALLTOALLV, and MPI_ALLTOALLW for intracommunicators.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2008-12-08 10:51:25 -0600**

---

Updated proposed solution to bring it up at the Forum

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2008-12-16 12:15:04 -0600**

---

Fixed typo in impact on implementations

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-10 08:45:35 -0600**

---

Updated proposal based on last forum feedback. Please review the proposal before the next meeting! 

Thanks, 
   Torsten

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-19 07:34:16 -0600**

---

Please add in your Solution:

  Add at the end of p156.2:

```
The "in place" option is not supported for intercommunicators.
```

Please add Change-Log entry:

```
Section 5.8 on page 155.[[BR]]
"In place" option is added to MPI_ALLTOALL for intracommunicators.
```

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-19 14:14:01 -0600**

---

Replying to RolfRabenseifner:

> Please add in your Solution:
> 
>   Add at the end of p156.2:
>     The "in place" option is not supported for intercommunicators.
> I disagree because p132:24 already restricts the in place option to intracommunicators. No collective operation says anything about in place and intracomms. Consider also p134:36.
> 
> Please add Change-Log entry:
> Section 5.8 on page 155.
> "In place" option is added to MPI_ALLTOALL for intracommunicators.
> done

Best,
  Torsten

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-22 05:11:50 -0600**

---

Minor change in ticket text, proposal reviewed, ok with me.

Do we need to do the trivial implementation? (that'll probably run almost as fast as many
MPI_Alltoall implementations, and already provide the space saving advantage). It's a 10-liner

Maybe consider Rationale for MPI_Alltoallv/w (p.157, line 20, p.159, line 17):

Supporting the "in place" option would imply that each pair of processes specifies the same amount of data (with the same type signature) to be sent and received. It was thought that this extra semantic constraint would lead too easily to programming errors, and this concern outweighted the potential benefits of supporting the "in place" option in this case.

Or, why not, support also MPI_IN_PLACE for MPI_Alltoallv and MPI_Alltoallw??

Jesper

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-22 11:52:06 -0600**

---

Thinking about it some more, I'd be quite in favor for allowing also MPI_IN_PLACE for MPI_Alltoallv/w. The alltoall exchangedimplied by giving "in place" on ALL processes is a "symmetric" alltoall exchange (process i sends and receives the same amount to process j, but amounts can be different for different j's). This is somewhat less irregular than the full-fledged alltoallv/w, and I think interesting algorithms exists than can give real benefits (see e.g.
Sanders/Traff, "Hierarchical Factor Algorithm"). Maybe there's even applications that make use of this kind of symmetric alltoall?

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-25 14:46:17 -0600**

---

Yes Torsten, you are right with rejecting my comment on 156.2.

New review: Currently proposed solution is okay (as far as I can see).

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-28 02:37:56 -0600**

---

Added advice to implementors, and Part B: MPI_IN_PLACE for MPI_ALLTOALLV/W, as discussed in teleconf 27.1.09.

Please review. Note also typos reported in ticket #113.

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-01-28 05:26:22 -0600**

---

Generally looks good. A comma is missing in the Part A advice to implementers: "(e.g. constant)" should be "(e.g., constant)". Also, an extra "than" appears in that advice: "than than" should be "than".

You could simplify the Part B advice to users by changing "Users must be careful to ensure that" to "Users must ensure that". Also, this sentence: "This symmetric exchange is useful in some applications where the data to be sent are not needed after the MPI_ALLTOALLV exchange." does not read quite right. The data is needed after the exchange at the receiver but not at the sender. It might be best to delete it; otherwise it would be good if you could make it closer to the Part A Rationale -- something like "This symmetric exchange is useful in some applications such as BLAH." 

You might want to update the description where it says "The idea could be extended to MPI_Alltoallv, but the requirements on the count vectors would probably make this too complicated for users?" since Part B now does that so the old comment is confusing.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-29 11:06:42 -0600**

---

Replying to bronis:

> Generally looks good. A comma is missing in the Part A advice to implementers: "(e.g. constant)" should be "(e.g., constant)". Also, an extra "than" appears in that advice: "than than" should be "than".
> 
> fixed
> 
> You could simplify the Part B advice to users by changing "Users must be careful to ensure that" to "Users must ensure that". Also, this sentence: "This symmetric exchange is useful in some applications where the data to be sent are not needed after the MPI_ALLTOALLV exchange." does not read quite right. The data is needed after the exchange at the receiver but not at the sender. It might be best to delete it; otherwise it would be good if you could make it closer to the Part A Rationale -- something like "This symmetric exchange is useful in some applications such as BLAH." 
> 
> also fixed. The advice to users now reads like the one for MPI_Alltoall. I agree that it might be
> good to cite an actual app
> 
> You might want to update the description where it says "The idea could be extended to MPI_Alltoallv, but the requirements on the count vectors would probably make this too complicated for users?" since Part B now does that so the old comment is confusing.
> 
> changed

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-01-29 15:00:07 -0600**

---

It's is almost set -- you missed my comment about the data being sent. Please change this sentence in the Part B advice to users:

This symmetric exchange is useful in some applications where the data to be sent will not be used after the MPI_Alltoallv exchange.

To:

This symmetric exchange is useful in some applications where the data to be sent will not be used by the sender after the MPI_Alltoallv exchange.

That makes it clear that the receiver still will use it.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-02-02 18:18:13 -0600**

---

Since this IN_PLACE is actually a REPLACE, the implementation must provide its own buffering for the data. As a result, the implementation of this functionality involves more work compared to other things being added to MPI 2.2 (which are more or less simple changes). I would prefer this being moved to 3.0.

Some comments on the proposed text:
- The rest of the collectives chapter says "data is taken from" rather than "data are taken from". I would prefer to use the same here. (Now let's not start a new ticket to change that everywhere :-).)
- "Data sent and received have the same type map..." should be "Data sent and received must have the same type map..." in all 3 places.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-02-04 03:09:29 -0600**

---

Bronis: I added "by the sending process" in both part A and B

Rajeev: I did the two updates. I do not think that implementing MPI_IN_PLACE is so much work that this is out of the scope for MPI 2.2

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-04 09:55:10 -0600**

---

Looks fine.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-04 17:08:31 -0600**

---

To match the text in the rest of the chapter, we should changes lines like the following:
"In that case the sendcount and sendtype arguments are ignored."
to:
"In such a case, sendcount and sendtype are ignored."
("that" --> "such a") and (drop "the" and "arguments")

Also, change statements like:
"is specified by passing the value MPI_IN_PLACE to the argument sendbuf at all processes."
to:
"is specified by passing MPI_IN_PLACE as the value of sendbuf at all processes.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-04 17:56:17 -0600**

---

The "same type map" statements are subtle:

"Data sent and received must have the same type map as specified by recvcount and recvtype."

It's easy for a reader to get this concept confused with "same type signature".  I'd recommend rephrasing it or calling attention to the distinction to make it clear to the reader.

For allgather, (p 154, line 26), we currently have:

"Then the input data of each process is assumed to be in the area where that process would receive its own contribution to the receive buffer."

We could say something similar for alltoall (with classic i/j language):

"Then the data in the j-th block sent from process i is assumed to be in the area where process i would receive its data from process j."

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-04 18:03:25 -0600**

---

Could use some re-wording here:

"An MPI implementation should strive to use significantly less (e.g., constant) extra memory than the size of the user provided receive buffer, and balance this against performance."

to:

"Users may opt to use the "in place" option in order to conserve memory.  Quality MPI implementations should strive to minimize system buffering in such cases."

Also remember to add the "(End of advice to implementors.)" and similar statements.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-04 18:27:00 -0600**

---

Perhaps it would help clarify things to add descriptions like the following since they exist in each of the Alltoall descriptions:

When using the "in place" option, the outcome is as if each process executed an MPI_SENDRECV_REPLACE with every other process with:

MPI_Sendrecv_replace(recvbuf + i_recvcount_extent(recvtype), recvcount, recvtype, i, ..., i, ...);

For alltoallv, this illustration would change to:

MPI_Sendrecv_replace(recvbuf + rdispls[i]*extent(recvtype), recvcounts[i], recvtype, i, ..., i, ...);

And for alltoallw:

MPI_Sendrecv_replace(recvbuf + rdispls[i], recvcounts[i], recvtype[i], i, ..., i, ...);

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-02-05 15:25:49 -0600**

---

Minor change request on the Change-Log entry:[[BR]]
Each section is cited only with its starting page.
Therefore, only page number 155 aplies. Please, can you reduce the entry to one line with all three routine names.

Major concern about this ticket:

Question to the MPI-Forum: 

  MPI-2.1 p132:22-27 tells: 

```
All processes in the group identified by the intracommunicator must call the
collective routine with matching arguments. 

In many cases, collective communication can occur “in place” for 
intracommunicators, with the output buffer being identical to the input buffer. 
This is specified by providing a special argument value, MPI_IN_PLACE, instead 
of the send buffer or the receive buffer argument, depending on the operation 
performed. 
```

  Does this mean that MPI_IN_PLACE must be given on all or none processes? 
  I'll put this question also on ticket #94 (MPI_IN_PLACE at MPI_EXSCAN). 

If the answer is "all processes", then this answer should be handled with
a comment on page 132 and the "all processes" in this proposal should be removed
or added to all locations of MPI_IN_PLACE.

If the answer is that MPI_IN_PLACE need not to match, then this should be also
told on page 132 and the "all processes" in this proposal makes sense.

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-05 16:33:25 -0600**

---

Answer to Rolf:

See my response to this question in ticket #94. MPI_IN_PLACE is not required everywhere. Noting that here would be fine although perhaps it should be clarified generally.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-08 07:43:51 -0600**

---

Reply to Adam:
1) simple wording changes for consistency are applied nowThe "same type map" statements are subtle: 
2) "The "same type map" statements are subtle:"
  It may be subtle but it's clear wording. I'd say that a standard document has to be clear, and I think the description is very clear. 
3) wording change about memory consumption applied with a small change (end of advices are automatically added by the latex environment, they can not be forgotten
4) proposal for additional clarification: 
  I think that the concept of "in place" is simple enough to not need additional clarification. It has no global impact and all previous i,j discussions apply to it with the difference that send/recv buffers are merged (which is stated in the introduction).

Rolf:
1) changelog fixed
2) agree to Bronis - it's inconsistent in the MPI-2.1 right now:
  Reduce/Gather(v)/Scatter(v) - root only
  Allreduce/Allgather(v) - all
  Reduce_scatter/Scan - undefined
  thus, "all processes" is consistent with the other "all"-operations, however, we might want to fix/clarify Reduce_scatter and Scan in MPI-3 (enforcing it on all procs might break user code)

Please review.

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-08 10:30:01 -0600**

---

By Adam's note, you need to delete "arguments" in the first proposed paragraph with "sendcount and sendtype arguments" and in the paragraph for MPI_ALLTOALLV with "sendcounts, sdispls and sendtype arguments" and in the paragraph for MPI_ALLTOALLW with "sdispls and sendtypes arguments".

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-08 11:29:52 -0600**

---

fixed - thanks Bronis!

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-08 13:58:24 -0600**

---

Looks good at this point to me.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-09 13:05:23 -0600**

---

Rajeev, Jesper, Adam, Rolf - please review/approve again!

Rajeev: I think those things should be fixed by the chapter authors, but we need a more efficient way to handle this. I propose we should have _one_ ticket per chapter that fixes typos and minor things (such as non-.+. vs. non.+, "void _" vs. "void_", or "data is" vs. "data are" ...).

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-02-09 13:26:51 -0600**

---

reviewed, and (almost) fine with me. We could replace " halves the memory consumption" by " halves the user memory consumption", which makes the advice to implementors even more meaningful.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-09 16:00:36 -0600**

---

A few minor edits:

1) need a comma after prepositional phrase: "In such a case"  -->  "In such a case,"

2) "some applications" sounds weaker than it could be stated

Change:

"This symmetric exchange is useful in some applications where the data to be sent will not be used by the sending process after the MPI_Alltoallv exchange."

To: (change "is" to "can be" and drop "some")

"This symmetric exchange can be useful in applications where the data to be sent will not be used by the sending process after the MPI_Alltoallv exchange."

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-09 16:01:44 -0600**

---

Also, why do we only include this comment for alltoallv:

"This symmetric exchange is useful in some applications where the data to be sent will not be used by the sending process after the MPI_Alltoallv exchange."

Doesn't this also apply to standard alltoall and alltoallw?

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-02-09 16:07:50 -0600**

---

Finally, leaving the "type map" language in place is fine.  It is certainly clear; it's just subtle and I just know people will confuse this with "type signature".

I'll still strongly argue that having the MPI_Sendrecv_replace() examples would be very helpful to illustrate the data movement to the reader (and make the above subtle point clearer).  I've restated the text here:

When using the "in place" option, the outcome is as if each process executed an MPI_SENDRECV_REPLACE with every other process with a call to:

MPI_Sendrecv_replace(recvbuf + i_recvcount_extent(recvtype), recvcount, recvtype, i, ..., i, ...);

For alltoallv, this illustration would change to:

MPI_Sendrecv_replace(recvbuf + rdispls[i]*extent(recvtype), recvcounts[i], recvtype, i, ..., i, ...);

And for alltoallw:

MPI_Sendrecv_replace(recvbuf + rdispls[i], recvcounts[i], recvtype[i], i, ..., i, ...);

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-02-09 17:37:24 -0600**

---

Hi Adam,

I corrected 1) and 2). For MPI_Alltoallw "in place" usage, there's now a reference to MPI_Alltoallv (and therefore no explicit advice to users). I think the explanation of the feature in terms of MPI_Sendrecv_replace is unnecessary.

For the other reviewers: is it OK with you?

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-09 17:53:36 -0600**

---

Looks fine.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-02-09 18:03:34 -0600**

---

A couple of instances of "data are taken" still remain. Last lines of new paras on 157:20 and 159:17 "and are taken from the locations of the receive buffer specified by rdispls."

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-09 18:49:19 -0600**

---

reviewed Jesper's changes and agree!

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-02-11 12:53:19 -0600**

---

We are awaiting an open source implementation for inspection.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-20 20:19:51 -0600**

---

**Attachment added:** [`patch_alltoall_in_place_ompi-1.3.diff`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/31/patch_alltoall_in_place_ompi-1.3.diff) (7.4 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-20 20:21:24 -0600**

---

implementation as patch vs. Open MPI 1.3 (basic coll component) is attached. This should meet all requirements (reviews are also complete).

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-06-18 13:07:29 -0500**

---

**Attachment added:** [`LICENSE`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/31/LICENSE) (1.9 KiB)
License of the attached code

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-06-18 13:08:27 -0500**

---

Attached BSD-style license just to be sure.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-07-27 15:06:23 -0500**

---

Moved to proper type

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-07-31 16:04:19 -0500**

---

**Attachment added:** [`ticket31-coll.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/31/ticket31-coll.pdf) (71.1 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-07-31 16:05:55 -0500**

---

Intentional edits:
- Used language neutral function names when referring to other functions in text, e.g., "MPI_Alltoallv" --> "MPI_ALLTOALLV".

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-03 12:47:45 -0500**

---

**Attachment added:** [`change-log_ticket31_item9.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/31/change-log_ticket31_item9.pdf) (424.5 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-08-12 02:01:14 -0500**

---

Reviewed, fine!

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-08-22 15:03:36 -0500**

---

PDF review - ok!

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-08-27 05:42:20 -0500**

---

[Reminder] please review!

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-08-27 11:40:49 -0500**

---

The fix to the text in the standard is ok.

A minor thing for the entry in the change log is the quotes before In place are not the right type.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-08-29 05:51:02 -0500**

---

[Reminder] please review!

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-29 13:59:06 -0500**

---

Corrected in svn r434: "In place" -> ``In place''

pdf not redone, but checked in full mpi_report.tex -> now okay.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-08-29 18:29:30 -0500**

---

Has four OK reviews (all complaints are now fixed) -> passed

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2010-09-18 05:00:13 -0500**

---

This ticket is (long-since) complete; marking it resolved/text committed.
